### PR TITLE
Example where `Context???` doesn't allow typical state management

### PR DESCRIPTION
**Pull request by ashpil**
Here it is! 

This is me trying to recreate a pattern I often see in `wgpu` examples. The idea is that you have a function that creates all of your required state and stores it in a struct, and then you have another function that renders one specific draw call, whereas your event loop remains in `main.`

This doesn't work with the current system, as you can't really store the necessary `Context???` type or use it in a function because of the `?`s. 

This currently doesn't compile.

### CODE REVIEW

File: examples/hello_triangle.rs [hunk 1]
```diff
@@ -11,145 +13,159 @@ use winit::{
...
+use std::error::Error;
+
+struct State {
+    queue: Queue,
+    device: Device,
+    swap_chain: SwapChain,
+    program: GraphicsProgram,
+    vertex_position: Vertex<Vec<[f32; 3]>>,
+    vertex_brightness: Vertex<Vec<f32>>,
+    context: WhatDoIPutHere,
```
Comment (ashpil) [inline examples/hello_triangle.rs:32]: Herein lies the issue.

**Comment by ashpil** [comment]
Of course one can create the `context` in each iteration, but doing unnecessary work in the loop sounds not very good.

Even beyond the struct issue at hand, it seems like one should be able to do with `context` the things one expects to be able to do with any type - i.e., pass it to a function, store it in a struct, etc. But one cannot do that now.

**Comment by Pat-Lafon** [comment]
Now that I'm looking at the code, I'm reminded of a sharp corner of the design that I've briefly discussed with @Checkmate50 and I've opened #19 to remind me of it. I think that issue hits on my general issue with putting a `Context???` struct in some kind of state struct unless you are also including your `RenderPass`. I'm not sure I've seen a program that sets up the `RenderPass` in state and I would argue that the design of `Context` is to be only used with `RenderPass`. These two structs represent the static environment of the pipeline and the runtime environment of the pipeline respectively.

There are two other points here as far as I can tell:
- Unnecessary work: I'm pretty confident that the "extra work" of creating an empty struct is not measurable, especially when compared against the cost of talking with the gpu. Maybe there's a small cost to readability? I'm not sure if an extra line is that bad but this is probably subjective.
- One should be able to use this type in functions/closures: This actually isn't that clear to me. The only thing you can use this type for is to bind parameters to a render pass. Is there an example of a program that gives a render/compute pass to a function or closure? Such an example would be more convincing. 

The last point is interesting, I might even want a programmer to not be able to pass a Context via a closure/function. It seems like it could be an easy way for the programmer to shoot themselves in the foot. Again, a program that passes a render or compute pass to a function would be very interesting to look at.

**Comment by ashpil** [comment]
> It seems like it could be an easy way for the programmer to shoot themselves in the foot.

@Pat-Lafon Why do you think that? I don't see a reason that would be the case.

**Comment by Pat-Lafon** [comment]
> > It seems like it could be an easy way for the programmer to shoot themselves in the foot.
> 
> @Pat-Lafon Why do you think that? I don't see a reason that would be the case.

Currently, `Context`'s are suppose to have a nested, tree-like structure which matches the underlying runtime with no additional cost. I would be worried that if you start storing `Context`'s in closures and functions that you could more easily violate this. I think the kind of feature your asking for requires strong justification or a compelling example that can't be done with the current library.

**Comment by Pat-Lafon** [comment]
As we discussed today, if you don't know the name of Context Type, then you can't create a partially bound context.

**Pull request closed without merging**